### PR TITLE
refactor(sdk): simplify InputSpec

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
Similar to #8415.

This removes some trickery in the definition and usage of `InputSpec` that was originally 
intended to make deserialization from v1 component YAML turnkey. The v1 backward compatibility 
handling we have added since them makes this no longer helpful and decreases the clarity of the 
code.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more 
about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/components/structures.py [hunk 1]
```diff
@@ -38,39 +37,21 @@
 
 
 @dataclasses.dataclass
-class InputSpec_:
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/structures.py:41]: Core changes are here. The rest is to update tests and usage in response to this refactoring.

File: sdk/python/kfp/components/structures.py [hunk 2]
```diff
@@ -135,6 +116,13 @@ def _validate_type(self) -> None:
         if not spec_type_is_parameter(self.type):
             type_utils.validate_bundled_artifact_type(self.type)
 
+    def _validate_usage_of_optional(self) -> None:
+        """Validates that the optional and default properties are in consistent
+        states."""
+        # Because None can be the default value, None cannot be used to to indicate no default. This is why we need the optional field. This check prevents users of InputSpec from setting these two values to an inconsistent state, forcing users of InputSpec to be explicit about optionality.
+        if self.optional is False and self.default is not None:
+            raise ValueError('')
```
[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply sdk/python/kfp/components/structures.py]: Fixed

File: sdk/python/kfp/components/structures.py [hunk 3]
```diff
@@ -38,39 +37,21 @@
...
-
-    @functools.wraps(InputSpec_.__init__)
-    def __init__(self, *args, **kwargs) -> None:
-        """InputSpec constructor, which can access the arguments passed to the
-        constructor for setting ._optional value."""
-        if args:
-            raise ValueError('InputSpec does not accept positional arguments.')
-        super().__init__(*args, **kwargs)
-        self._optional = 'default' in kwargs
+    optional: Optional[bool] = False
```
Comment (chensun) [inline sdk/python/kfp/components/structures.py:50]: nit: looks like `optional` will never be `None`? If so, we can remove `Optional[...]` in the type annotation.

**Review by chensun — COMMENTED** [review]
/lgtm

Thanks!

**Comment by connor-mccarthy** [comment]
[Test failures](https://github.com/kubeflow/pipelines/pull/8422#issuecomment-1312214289) are unrelated to these changes

**Comment by connor-mccarthy** [comment]
@chensun, good catch. Fixed by https://github.com/GoogleCloudPlatform/oss-test-infra/pull/1835.

**Pull request merged**